app/javascript/lib/validation/expression-matrices-validation.js

```diff
@@ -0,0 +1,263 @@
+} from './shared-validation'
+
+/** Parse a dense matrix file */
+export async function parseDenseMatrixFile(chunker, mimeType, fileOptions) {
+  const { headers, delimiter } = await getParsedHeaderLines(chunker, mimeType)
+  let issues = validateDenseHeader(headers)
+
+  const dataObj = {} // object to track multi-line validation concerns
+  await chunker.iterateLines((rawLine, lineNum, isLastLine) => {
+    const line = parseLine(rawLine, delimiter)
+    issues = issues.concat(validateValuesAreNumeric(line, isLastLine, lineNum, dataObj))
+    issues = issues.concat(validateColumnNumber(line, isLastLine, headers, lineNum, dataObj))
+    issues = issues.concat(validateUniqueCellNamesWithinFile(line, isLastLine, dataObj))
+    issues = issues.concat(validateMetadataLabelMatches(headers, line, isLastLine, dataObj))
+    issues = issues.concat(validateGroupColumnCounts(headers, line, isLastLine, dataObj))
+    // add other line-by-line validations here
+  })
+  return { issues, delimiter, numColumns: headers[0].length }
+}
+
+/** Parse an MTX matrix file */
+export async function parseSparseMatrixFile(chunker, mimeType, fileOptions) {
+  let issues = []
+  const delimiter = '  ' // tab delimited
+  const dataObj = {} // object to track multi-line validation concerns
+  await chunker.iterateLines((rawLine, lineNum, isLastLine) => {
+    issues = issues.concat(validateMTXHeaderLine(rawLine, lineNum, dataObj))
+    issues = issues.concat(validateSparseColumnNumber(rawLine, isLastLine, lineNum, dataObj))
+    issues = issues.concat(validateSparseNoBlankLines(rawLine, isLastLine, lineNum, dataObj))
+    // add other line-by-line validations here
+  })
+  return { issues, delimiter, numColumns: dataObj.correctNumberOfColumns }
+}
+
+
+/** Parse a barcodes file, and return an array of issues, along with file parsing info */
+export async function parseBarcodesFile(chunker, mimeType, fileOptions) {
+  let issues = []
+
+  const dataObj = {} // object to track multi-line validation concerns
+  await chunker.iterateLines((rawLine, lineNum, isLastLine) => {
+    issues = issues.concat(validateUniqueRowValuesWithinFile(rawLine, isLastLine, dataObj))
+  })
+  return { issues }
+}
+
+
+/** Parse a features file, and return an array of issues, along with file parsing info */
+export async function parseFeaturesFile(chunker, mimeType, fileOptions) {
+  let issues = []
+
+  const dataObj = {} // object to track multi-line validation concerns
+  await chunker.iterateLines((rawLine, lineNum, isLastLine) => {
+    issues = issues.concat(validateUniqueRowValuesWithinFile(rawLine, isLastLine, dataObj))
+  })
+  return { issues }
+}
+
+/**
+ * Verify cap format for an expression matrix file
+ *
+ * The "cap" for an expression matrix file is the first row
+ */
+function validateDenseHeader([headers]) {
+  let issues = []
+  if (!headers) {
+    return [['error', 'format:cap:no-cap-row', 'File does not have a non-empty header row']]
```

Comment:
Less redundant and a bit clearer:

```suggestion
    return [['error', 'format:cap:no-header-row', 'File does not have a non-empty header row']]
```